File: python/src/celladmix/dataset.py
"""Dataset-level Python API."""

from __future__ import annotations

import hashlib
import os
import shutil
from pathlib import Path
from typing import Optional

from . import _core
from .fit import CellAdmixFit
from .io import annotation_vectors, ensure_list, read_annotation, recommended_rank
from .state import clustering_result_to_frame


def _default_threads() -> int:
    return max(1, min(10, os.cpu_count() or 1))


def _annotation_hash(annotation) -> str:
    if annotation is None:
        return ""
    digest = hashlib.sha1()
    for cell, label in sorted(zip(annotation.index.astype(str), annotation.astype(str))):
        digest.update(cell.encode())
        digest.update(b"=")
        digest.update(label.encode())
        digest.update(b";")
    return digest.hexdigest()


_COMPARABLE_FIT_PARAMS = (
    "rank", "nmf_variant", "nmf_init", "molecule_scoring", "ncv_k", "graph_k",
    "same_label_ratio", "nmf_iterations", "nmf_n_runs", "nmf_train_max_rows",
    "nmf_min_molecules", "seed", "training_scope_cell_types",
)


def _fit_param_diff(manifest: dict, requested: dict, annotation_hash: str) -> list[str]:
    """Describe requested fit parameters that differ from a cached run.

    Only caller-specified parameters participate; automatically resolved
    values (e.g. an omitted ncv_k) match whatever the cached run recorded.
    """

    def normalize(value):
        if isinstance(value, (list, tuple, set)):
            return ",".join(sorted(map(str, value)))
        return str(value)

    recorded_options = manifest.get("pipeline_options", {}) or {}
    diff = []
    for name in _COMPARABLE_FIT_PARAMS:
        if name not in requested or name not in recorded_options:
            continue
        recorded = normalize(recorded_options[name])
        wanted = normalize(requested[name])
        if recorded != wanted:
            diff.append(f"{name}: {recorded} -> {wanted}")
    recorded_hash = manifest.get("annotation_hash") or ""
    if recorded_hash and annotation_hash and recorded_hash != annotation_hash:
        diff.append("annotation content")
    return diff


class CellAdmix:
    """A cellAdmix dataset backed by a Xenium bundle and an on-disk cache."""

    def __init__(
        self,
        source=None,
        *,
        output_dir: str | os.PathLike = "out",
        format: str = "xenium",
        annotation=None,
        annotation_col: Optional[str] = None,
        cell_id_col: str = "cell_id",
        num_threads: Optional[int] = None,
        cell_filter=None,
        gene_filter=None,
        min_qv: float = -1.0,
        keep_unassigned: bool = False,
        keep_non_gene: bool = False,
    ):
        self.source = None if source is None else Path(source)
        self.output_dir = Path(output_dir)
        self.format = format
        self.num_threads = int(num_threads or _default_threads())
        self.cell_filter = ensure_list(cell_filter)
        self.gene_filter = ensure_list(gene_filter)
        self.min_qv = float(min_qv)
        self.keep_unassigned = bool(keep_unassigned)
        self.keep_non_gene = bool(keep_non_gene)
        self.annotation = (
            read_annotation(annotation, annotation_col=annotation_col, cell_id_col=cell_id_col)
            if annotation is not None
            else None
        )
        self.store_dir = self.output_dir / "input_store"
        self.runs_dir = self.output_dir / "runs"
        self.output_dir.mkdir(parents=True, exist_ok=True)
        self.runs_dir.mkdir(parents=True, exist_ok=True)

    def __repr__(self) -> str:
        ann = "none" if self.annotation is None else f"{self.annotation.nunique()} labels"
        source = None if self.source is None else str(self.source)
        return (
            f"CellAdmix(format={self.format!r}, source={source!r}, "
            f"output_dir={str(self.output_dir)!r}, annotation={ann}, "
            f"num_threads={self.num_threads})"
        )

    @classmethod
    def attach_existing(
        cls,
        output_dir: str | os.PathLike,
        *,
        source=None,
        format: Optional[str] = None,
        annotation=None,
        annotation_col: Optional[str] = None,
        cell_id_col: str = "cell_id",
        num_threads: Optional[int] = None,
    ):
        """Attach to an existing cellAdmix output directory.

        This does not build or scan the input store. If ``source`` is omitted,
        the method uses the source path recorded in ``input_store/store.json``
        when available. Bundle-dependent methods such as membrane image
        discovery need a usable source path.
        """
        output_dir = Path(output_dir)
        store_dir = output_dir / "input_store"
        store_manifest = None
        if store_dir.exists():
            try:
                store_manifest = _core.read_input_store_manifest(str(store_dir))
            except Exception:
                store_manifest = None
        if source is None and store_manifest is not None:
            source = store_manifest.get("source_path") or None
        if format is None and store_manifest is not None:
            format = store_manifest.get("source_type") or "xenium"
        return cls(
            source,
            output_dir=output_dir,
            format=format or "xenium",
            annotation=annotation,
            annotation_col=annotation_col,
            cell_id_col=cell_id_col,
            num_threads=num_threads,
        )

    def load_fit(self, run_id_or_path) -> CellAdmixFit:
        """Load a persisted fit by run id or path."""
        run_path = Path(run_id_or_path)
        if not run_path.exists():
            run_path = self.runs_dir / str(run_id_or_path)
        manifest = _core.read_run_manifest(str(run_path))
        run_dir = Path(manifest["paths"]["root_dir"])
        return CellAdmixFit(self, str(run_dir), manifest)

    def set_annotation(self, annotation, *, annotation_col: Optional[str] = None, cell_id_col: str = "cell_id"):
        """Set or replace the active cell annotation."""
        self.annotation = read_annotation(annotation, annotation_col=annotation_col, cell_id_col=cell_id_col)
        return self

    def ensure_store(self, *, force: bool = False, materialize_molecules: bool = True, verbose: bool = True):
        """Build or reuse the on-disk input store."""
        if self.format != "xenium":
            raise NotImplementedError("Python v1 currently implements Xenium stores")
        if self.source is None:
            raise ValueError(
                "Cannot build an input store without a source path. "
                "Pass source=... or attach to an existing completed run."
            )
        # The C++ store builder owns parsing and parquet materialization; the
        # Python layer only normalizes user options and paths.
        return _core.build_xenium_store(
            str(self.source),
            str(self.store_dir),
            cell_filter=self.cell_filter,
            gene_filter=self.gene_filter,
            min_qv=self.min_qv,
            keep_unassigned=self.keep_unassigned,
            keep_non_gene=self.keep_non_gene,
            materialize_molecules=materialize_molecules,
            force=force,
            num_threads=self.num_threads,
        )

    def fit(
        self,
        *,
        rank: Optional[int] = None,
        rank_multiplier: float = 1.2,
        rank_cap: int = 30,
        nmf_variant: str = "ls_nmf",
        nmf_init: str = "auto",
        nmf_n_runs: Optional[int] = None,
        molecule_scoring: str = "gene_loadings",
        run_id: Optional[str] = None,
        overwrite: bool = False,
        num_threads: Optional[int] = None,
        verbose: bool = True,
        **kwargs,
    ) -> CellAdmixFit:
        """Fit NMF factors and molecule labels.

        By default, rank is inferred from the active annotation and independent
        NMF restarts use the dataset thread count.
        """
        self.ensure_store(force=False)
        threads = int(num_threads or self.num_threads)
        explicit_n_runs = nmf_n_runs
        rank = int(rank or recommended_rank(self.annotation, multiplier=rank_multiplier, cap=rank_cap))
        nmf_n_runs = int(nmf_n_runs or threads)
        run_id = run_id or f"fit_rank{rank}_{nmf_variant}"
        run_dir = self.runs_dir / run_id
        annotation_hash = _annotation_hash(self.annotation)
        if (run_dir / "run.json").exists():
            if overwrite:
                shutil.rmtree(run_dir)
            else:
                manifest = _core.read_run_manifest(str(run_dir))
                requested = dict(kwargs)
                requested.update(
                    rank=rank, nmf_variant=nmf_variant, nmf_init=nmf_init,
                    molecule_scoring=molecule_scoring)
                if explicit_n_runs is not None:
                    requested["nmf_n_runs"] = nmf_n_runs
                diff = _fit_param_diff(manifest, requested, annotation_hash)
                if not diff:
                    # Cached runs are reused only when the request matches them.
                    if verbose:
                        recorded = manifest.get("package_version") or ""
                        note = (
                            f"; fitted with package {recorded}, pass overwrite=True to refit"
                            if recorded and recorded != _core.core_version()
                            else ""
                        )
                        print(f"Reusing cached run {run_id} (parameters match{note})")
                    return CellAdmixFit(self, str(run_dir), manifest)
                if verbose:
                    print(f"Parameters changed for run {run_id} ({', '.join(diff)}); refitting")
                shutil.rmtree(run_dir)

        cell_ids, labels = annotation_vectors(self.annotation)
        # Training strata are aligned to the store cell order inside C++.
        training_strata = _core.training_strata_for_cells(str(self.store_dir), cell_ids, labels)
        result = _core.fit_store(
            str(self.store_dir),
            str(run_dir),
            rank,
            nmf_variant=nmf_variant,
            nmf_init=nmf_init,
            molecule_scoring=molecule_scoring,
            nmf_n_runs=nmf_n_runs,
            num_threads=threads,
            training_cell_strata=training_strata,
            annotation_hash=annotation_hash,
            verbose=verbose,
            **kwargs,
        )
        explicit_ncv_k = kwargs.get("ncv_k")
        if explicit_ncv_k and int(explicit_ncv_k) > 0:
            n_genes = len(result.get("genes", []))
            if n_genes > 100 * int(explicit_ncv_k):
                import math
                import warnings

                warnings.warn(
                    f"ncv_k={explicit_ncv_k} is small for a {n_genes}-gene panel: "
                    "neighborhoods carry almost no gene co-occurrence signal and "
                    "NMF factors become seed-dependent. Consider the automatic "
                    f"default (omit ncv_k; ~{round(20 * math.sqrt(n_genes / 400))} "
                    "for this panel, subject to cell size) or gene subsetting.",
                    stacklevel=2,
                )
        return CellAdmixFit(self, str(run_dir), result)

    def cell_state_umap(
        self,
        *,
        cells_max: int | None = 5000,
        min_molecules: int = 10,
        min_genes: int = 5,
        n_variable_genes: int = 1000,
        pca_dims: int = 30,
        graph_k: int = 15,
        cluster_resolution: float = 1.0,
        compute_umap: bool = True,
        umap_neighbors: int = 15,
        umap_epochs: int = 200,
        umap_parallel_optimization: bool = True,
        normalization_scale: float = 5000.0,
        seed: int = 1,
        num_threads: Optional[int] = None,
    ):
        """Compute a cell-state UMAP from original input-store counts."""
        self.ensure_store(force=False, materialize_molecules=False)
        result = _core.cluster_store_counts(
            str(self.store_dir),
            min_molecules=min_molecules,
            min_genes=min_genes,
            cells_max=-1 if cells_max is None else int(cells_max),
            n_variable_genes=n_variable_genes,
            pca_dims=pca_dims,
            graph_k=graph_k,
            cluster_resolution=cluster_resolution,
            compute_umap=compute_umap,
            umap_neighbors=umap_neighbors,
            umap_epochs=umap_epochs,
            num_threads=int(num_threads or self.num_threads),
            umap_parallel_optimization=umap_parallel_optimization,
            normalization_scale=normalization_scale,
            seed=int(seed),
        )
        return clustering_result_to_frame(result, annotation=self.annotation)

File: r/R/dataset.R
#' High-Level cellAdmix Dataset
#'
#' R6 object storing source metadata, active annotations, fitted runs, scores,
#' and corrections. Construct with [cellAdmix()].
#' @export
CellAdmixDataset <- R6::R6Class(
  "CellAdmixDataset",
  public = list(
    prep = NULL,
    format = NULL,
    output_dir = NULL,
    source_info = NULL,
    active_annotation = NULL,
    active_domain = NULL,
    fits = NULL,
    num_threads = NULL,

    initialize = function(prep, format, source_info = NULL, annotation = NULL,
                          annotation_name = "manual", annotation_col = NULL,
                          cell_id_col = "cell_id", num_threads = NULL) {
      self$prep <- prep
      self$format <- format
      self$output_dir <- prep$project_dir
      self$source_info <- source_info %||% prep$source
      self$num_threads <- as.integer(num_threads %||% .celladmix_default_threads())
      self$fits <- list()
      private$.annotations <- list()
      private$.domains <- list()
      if (!is.null(annotation)) {
        self$set_annotation(
          annotation,
          name = annotation_name,
          annotation_col = annotation_col,
          cell_id_col = cell_id_col,
          active = TRUE
        )
      }
    },

    print = function(...) {
      file_label <- function(path) {
        if (is.null(path) || !length(path) || is.na(path[[1]]) || !nzchar(path[[1]])) {
          return(NULL)
        }
        basename(path[[1]])
      }
      cat("cellAdmix dataset\n")
      cat("  format:", self$format, "\n")
      cat("  output cache: configured\n")
      cat("  threads:", self$num_threads, "\n")
      probe <- self$prep$probe %||% list()
      if (!is.null(probe$transcripts$preferred_path)) {
        cat("  molecules:", file_label(probe$transcripts$preferred_path), "\n")
      } else if (!is.null(probe$molecules$path)) {
        cat("  molecules:", file_label(probe$molecules$path), "\n")
      }
      if (!is.null(probe$cells$preferred_path) && nzchar(probe$cells$preferred_path)) {
        cat("  cells:", file_label(probe$cells$preferred_path), "\n")
      }
      anns <- self$annotations()
      if (nrow(anns)) {
        active <- self$active_annotation %||% ""
        active_row <- match(active, anns$name)
        cat("  annotation:", active, " (", anns$n_cells[active_row],
            " cells, ", anns$n_labels[active_row], " labels)\n", sep = "")
      } else {
        cat("  annotation: none\n")
        cat("  note: provide annotation=, call $set_annotation(), or run $cluster()\n")
      }
      invisible(self)
    },

    summary = function() {
      list(
        format = self$format,
        output_dir = self$output_dir,
        source = self$source_info,
        annotations = self$annotations(),
        active_annotation = self$active_annotation,
        num_threads = self$num_threads,
        domains = self$domains(),
        active_domain = self$active_domain,
        fits = names(self$fits)
      )
    },

    annotations = function() {
      if (!length(private$.annotations)) {
        return(data.frame(name = character(), active = logical(), n_cells = integer(),
          n_labels = integer(), hash = character(), stringsAsFactors = FALSE))
      }
      out <- do.call(rbind, lapply(names(private$.annotations), function(name) {
        ann <- private$.annotations[[name]]
        data.frame(
          name = name,
          active = identical(name, self$active_annotation),
          n_cells = ann$n_cells,
          n_labels = ann$n_labels,
          hash = ann$hash,
          path = ann$path,
          stringsAsFactors = FALSE
        )
      }))
      rownames(out) <- NULL
      out
    },

    annotation = function(name = NULL, as_vector = FALSE) {
      name <- name %||% self$active_annotation
      if (is.null(name) || !(name %in% names(private$.annotations))) {
        stop("Unknown annotation: ", name %||% "<none>")
      }
      ann <- private$.annotations[[name]]
      if (isTRUE(as_vector)) ann$labels else ann$table
    },

    set_annotation = function(annotation, name = NULL, active = TRUE,
                              annotation_col = NULL, cell_id_col = "cell_id") {
      name <- .celladmix_clean_name(name %||% if (!length(private$.annotations)) "manual" else "annotation",
        "annotation")
      ann <- .celladmix_make_annotation(
        annotation,
        annotations_dir = self$prep$paths$annotations_dir,
        name = name,
        cell_id_col = cell_id_col,
        label_col = annotation_col
      )
      private$.annotations[[ann$name]] <- ann
      if (isTRUE(active) || is.null(self$active_annotation)) {
        self$active_annotation <- ann$name
      }
      invisible(ann)
    },

    use_annotation = function(name) {
      if (!(name %in% names(private$.annotations))) {
        stop("Unknown annotation: ", name)
      }
      self$active_annotation <- name
      invisible(self)
    },

    drop_annotation = function(name) {
      private$.annotations[[name]] <- NULL
      if (identical(self$active_annotation, name)) {
        self$active_annotation <- names(private$.annotations)[[1]] %||% NULL
      }
      invisible(self)
    },

    ensure_store = function(required = c("molecules", "counts"), ...) {
      required <- match.arg(required)
      dots <- .celladmix_dots_with_default_threads(list(...), self$num_threads)
      do.call(.celladmix_as_store, c(list(self$prep, required = required), dots))
    },

    store = function(required = c("molecules", "counts"), ...) {
      self$ensure_store(required = required, ...)
    },

    counts = function(...) {
      dots <- .celladmix_dots_with_default_threads(list(...), self$num_threads)
      do.call(celladmix_collect_store_counts, c(list(self$prep), dots))
    },

    cell_state_umap = function(annotation = NULL, cells_max = 5000L,
                               min_molecules = 10L, min_genes = 5L,
                               n_variable_genes = 1000L, pca_dims = 30L,
                               graph_k = 15L, cluster_resolution = 1,
                               compute_umap = TRUE, umap_neighbors = 15L,
                               umap_epochs = 200L,
                               umap_parallel_optimization = TRUE,
                               normalization_scale = 5000, seed = 1L,
                               verbose = FALSE, ...) {
      ann <- private$resolve_annotation(annotation, register = TRUE)
      dots <- .celladmix_dots_with_default_threads(list(...), self$num_threads)
      store <- self$ensure_store(required = "counts",
        store_mode = "counts", num_threads = dots$num_threads,
        verbose = verbose)
      result <- .celladmix_cluster_store(
        store$path,
        min_molecules = min_molecules,
        min_genes = min_genes,
        cells_max = if (is.null(cells_max)) NA_integer_ else as.integer(cells_max),
        n_variable_genes = n_variable_genes,
        pca_dims = pca_dims,
        graph_k = graph_k,
        cluster_resolution = cluster_resolution,
        compute_umap = compute_umap,
        umap_neighbors = umap_neighbors,
        umap_epochs = umap_epochs,
        num_threads = dots$num_threads,
        umap_parallel_optimization = umap_parallel_optimization,
        normalization_scale = normalization_scale,
        seed = seed,
        clusters_out = tempfile("celladmix_state_clusters_", fileext = ".parquet"),
        embedding_out = tempfile("celladmix_state_embedding_", fileext = ".parquet"),
        verbose = verbose
      )
      .celladmix_clustering_frame(result, annotation = ann$labels)
    },

    cluster = function(name = "louvain", active = NULL, ...) {
      dots <- .celladmix_dots_with_default_threads(list(...), self$num_threads)
      clust <- do.call(celladmix_cluster_cells, c(list(self$prep, cluster_id = name), dots))
      labels <- stats::setNames(paste0("cluster_", clust$clusters$cluster), clust$clusters$cell_id)
      should_activate <- if (is.null(active)) is.null(self$active_annotation) else isTRUE(active)
      self$set_annotation(labels, name = name, active = should_activate)
      clust
    },

    identify_domains = function(name = NULL, annotation = NULL, active = TRUE, ...) {
      ann <- private$resolve_annotation(annotation, register = TRUE)
      domain_id <- .celladmix_clean_name(name %||% paste0(ann$name, "_domains"), "domains")
      dots <- .celladmix_dots_with_default_threads(list(...), self$num_threads)
      domains <- do.call(celladmix_identify_domains, c(list(
        prep = self$prep,
        annotations = ann$table,
        annotation_col = "label",
        annotation_cell_id_col = "cell_id",
        domain_id = domain_id
      ), dots))
      private$.domains[[domain_id]] <- domains
      if (isTRUE(active) || is.null(self$active_domain)) {
        self$active_domain <- domain_id
      }
      domains
    },

    load_domains = function(name) {
      domain_id <- .celladmix_clean_name(name, "domains")
      domains <- celladmix_load_domains(self$prep, domain_id)
      private$.domains[[domain_id]] <- domains
      if (is.null(self$active_domain)) {
        self$active_domain <- domain_id
      }
      domains
    },

    domains = function(name = NULL) {
      if (is.null(name)) {
        if (!length(private$.domains)) {
          return(data.frame(name = character(), active = logical(),
            n_cells = integer(), n_domains = integer(), stringsAsFactors = FALSE))
        }
        out <- do.call(rbind, lapply(names(private$.domains), function(domain_id) {
          domains <- private$.domains[[domain_id]]
          data.frame(
            name = domain_id,
            active = identical(domain_id, self$active_domain),
            n_cells = nrow(domains$domains %||% data.frame()),
            n_domains = length(unique((domains$domains %||% data.frame())$domain_label)),
            stringsAsFactors = FALSE
          )
        }))
        rownames(out) <- NULL
        return(out)
      }
      domain_id <- .celladmix_clean_name(name, "domains")
      private$.domains[[domain_id]] %||% stop("Unknown domains: ", domain_id)
    },

    list_fits = function() {
      data.frame(
        run_id = names(self$fits),
        run_dir = vapply(self$fits, function(x) x$run_dir, character(1)),
        annotation = vapply(self$fits, function(x) x$annotation_name, character(1)),
        rank = vapply(self$fits, function(x) as.integer(x$rank), integer(1)),
        stringsAsFactors = FALSE
      )
    },

    get_fit = function(run_id) {
      self$fits[[run_id]] %||% stop("Unknown fit: ", run_id)
    },

    read_fit = function(run_dir, annotation = NULL, name = NULL) {
      ann <- private$resolve_annotation(annotation, register = TRUE)
      run <- celladmix_read_run(run_dir)
      rank <- as.integer(run$n_factors %||% nrow(run$h) %||% NA_integer_)
      fit <- CellAdmixFit$new(
        dataset = self,
        run = run,
        annotation = ann,
        rank = rank,
        params = list(read_existing = TRUE)
      )
      key <- .celladmix_clean_name(name %||% basename(run$path), "fit")
      self$fits[[key]] <- fit
      fit
    },

    fit = function(annotation = NULL, rank = "auto", rank_multiplier = 1.2,
                   rank_cap = 30L, run_id = NULL, overwrite = FALSE, ...) {
      ann <- private$resolve_annotation(annotation, register = TRUE)
      resolved_rank <- if (is.null(rank) || identical(rank, "auto")) {
        .celladmix_annotation_rank(ann, multiplier = rank_multiplier, cap = rank_cap)
      } else {
        as.integer(rank[[1]])
      }
      dots <- list(...)
      dots <- .celladmix_dots_with_default_threads(dots, self$num_threads)
      nmf_variant <- dots$nmf_variant %||% "invsqrt_kl"
      run_id <- run_id %||% .celladmix_run_id("fit", ann, resolved_rank, nmf_variant)
      run_dir <- file.path(self$prep$paths$runs_dir, run_id)
      if (dir.exists(run_dir) && file.exists(file.path(run_dir, "run.json")) && !isTRUE(overwrite)) {
        fit <- self$read_fit(run_dir, annotation = ann, name = run_id)
        fit$params <- c(fit$params, list(reused_existing = TRUE, requested_rank = resolved_rank))
        return(fit)
      }
      run <- do.call(celladmix_fit, c(
        list(
          prep = self$prep,
          training_labels = ann,
          rank = resolved_rank,
          run_id = run_id
        ),
        dots
      ))
      explicit_ncv_k <- dots$ncv_k
      if (!is.null(explicit_ncv_k) && is.finite(explicit_ncv_k[[1]]) &&
          explicit_ncv_k[[1]] > 0) {
        n_genes <- length(run$genes)
        if (n_genes > 100 * as.integer(explicit_ncv_k[[1]])) {
          warning(sprintf(paste0(
            "ncv_k=%d is small for a %d-gene panel: neighborhoods carry ",
            "almost no gene co-occurrence signal and NMF factors become ",
            "seed-dependent. Consider the automatic default (omit ncv_k; ",
            "~%d for this panel, subject to cell size) or gene subsetting."),
            as.integer(explicit_ncv_k[[1]]), n_genes,
            as.integer(round(20 * sqrt(n_genes / 400)))), call. = FALSE)
        }
      }
      fit <- CellAdmixFit$new(
        dataset = self,
        run = run,
        annotation = ann,
        rank = resolved_rank,
        params = c(list(rank_policy = "auto_annotation", rank_multiplier = rank_multiplier,
          rank_cap = rank_cap, overwrite = isTRUE(overwrite)), dots)
      )
      self$fits[[run_id]] <- fit
      fit$save_metadata()
      fit
    }
  ),
  private = list(
    .annotations = NULL,
    .domains = NULL,

    resolve_annotation = function(annotation = NULL, register = FALSE) {
      if (is.null(annotation)) {
        if (is.null(self$active_annotation)) {
          stop("No active annotation is available. Provide annotation=, call $set_annotation(), or run $cluster().")
        }
        return(private$.annotations[[self$active_annotation]])
      }
      if (inherits(annotation, "celladmix_annotation")) {
        return(annotation)
      }
      if (is.character(annotation) && length(annotation) == 1L &&
          annotation %in% names(private$.annotations)) {
        return(private$.annotations[[annotation]])
      }
      if (!isTRUE(register)) {
        return(.celladmix_make_annotation(
          annotation,
          annotations_dir = self$prep$paths$annotations_dir,
          name = "override"
        ))
      }
      name <- if (is.character(annotation) && length(annotation) == 1L && !file.exists(annotation)) {
        annotation
      } else {
        paste0("annotation_", length(private$.annotations) + 1L)
      }
      self$set_annotation(annotation, name = name, active = FALSE)
      private$.annotations[[.celladmix_clean_name(name, "annotation")]]
    }
  )
)
